README.md
# AtlaNET

### About
AtlaNet is a network of systems working together to retrieve, store and provide P2000 alerts.\
\
![Example](http://atlanet.jkctech.nl/images/screenshots/github.png)

### Honorable Mentions
 - [1 FLEX Radio Tutorial | Richard IJzermans](https://raspberrytips.nl/p2000-meldingen-ontvangen/)
 - [2 Starter Code | oneguyoneblog](https://nl.oneguyoneblog.com/2016/08/09/p2000-ontvangen-decoderen-raspberry-pi/)
 - [3 Capcode Database Provider | Tom_Zulu](https://www.tomzulu10capcodes.nl/)

## Requirements
### Hardware
 - RTL2832U USB TV / Radio receiver (Aliexpress +- €10,-)
 - Raspberry Pi (Or some other Linux capable computer with USB)
 - Strobe Light (Optional + Requires soldering skills)
 - Arduino Nano (Optional for controlling the strobe light)


### Software
 - Python 2.7 (Specifically 2.7)
 - Pip (See requirements.txt for all the packages)
 - alsa alsa-tools (For volume management)
 - espeak (For the talking)
 - cmake build-essential libusb-1.0 qt4-qmake libpulse-dev libx11-dev qt4-default (For the radio, see link 1 in "Honorable Mentions")

Changelog.txt
11-09-2019
 - "Fixed"? bug where some messages are lost due to the main loop delay
   ^ Subject actively watched for next week to confirm a "fix"
 - Added an error logger function
 - Changed around some strings
 - Note: For the 2 people actually reading this repo, Processor & Twitter utils are planned to be moved to webserver instead of local client.

08-09-2019
 - This time I REALLY fixed issue #1 by rewriting the radio logic
 - Removed crash catcher because it's annoying me
 - AGAIN fixed encoding errors (Should be fine now)

06-09-2019
 - Fixed issue #1 (Double entries conflict).
 - Fixed bug where capcode description starting with number got interpreted as int.
 - Added support for multiple Twitter Hooks
 - Utilized Twitter hook and assigned special cases
 - Introduced dedicated area for processing a mnessage so the workflow can be defined.
 - Removed "Hide Pygame hello message".
 - hasTrigger() now returns the triggered string or capcode
 - Fixed some unicode bugs
 - Fixed priority lagging 1 alarm behind

04-09-2019
 - Added Twitter hook

03-09-2019
 - Fixed leaving radio running on crash
 - Program will now terminate other claiming threads when needed

02-09-2019
 - Added realtime capcode resolving from central database
 - Split some code up to multiple files
 - Replaced old regex prio patterns with new ones
 - Added getPrio()
 - Monitor now uses getPrio()
 - Added discipline identifiers to the monitor

01-09-2019
 - Fixed monitor taking up 100% CPU (oops)
 - Changed colors in documentation
 - Added capcode & Word filter to trigger alarms

30-08-2019
 - Added serial controller to turn on a strobe light when an alert is received

28-08-2019
 - Fixed uploading the wrong file...
 - Changed folder structure
 - Added documentation

26-08-2019
 - Added logging of raw messages and unique messages
 - Added ascii art header file
 - Added shortcut scripts for infrastructure
 - Minor logic fixes and changes
 - Added MySQL connector

23-08-2019
 - Fixed speaker speed
 - Added quick debug speaker to custom need
 - Added 2 quick ways to save data to a log file

18-08-2019
 - Added text to speach (TTS) controller

08-08-2019
 - Minor changes to the monitor script
 - Introduction of the filter script (will replace parts of the code in monitor.py)
 - Preparations for planned features including logging, caching and saving.

01-08-2019
 - Renamed p2000.py > monitor.py
 - Added config files (Rename config_default.json to config.json for proper usage)
 - Created web connector to send the messages to the server
 - Removed error log file code entirely

31-07-2019
 - Initial creation / adaptation of original code


src/monitor/monitor.py
#!/usr/bin/python

#
# Monitor.py
#
# AtlaNET P2000 Receiver - By: JKCTech
# https://github.com/jkctech/AtlaNET
#
# Inspired by: https://nl.oneguyoneblog.com/2016/08/09/p2000-ontvangen-decoderen-raspberry-pi/
# Follow up on a tutorial setting up the FLEX radio: https://raspberrytips.nl/p2000-meldingen-ontvangen/
#

import os
import time
import sys
import subprocess
import re
import fcntl
import requests
import mysql.connector
from termcolor import colored

# Import our own stuff <3
from utils.radio import *
from utils.settings import *
from utils.header import *
from utils.logger import *
from utils.alerter import *
from utils.filter import *
from utils.resolver import *
from utils.processor import *

# Set proper encoding
reload(sys)
sys.setdefaultencoding('utf8')

# Print header design
printheader()

# Get settings from files
settings = getSettings()

# Add triggers to settings
settings = setTriggers(settings)

# Download info from server
refreshServerLists(settings)

# Connect to database
con = None
if settings['mysql']['enabled']:
	print 'Database Connection:',
	try:
		con = mysql.connector.connect(
			host=settings['mysql']['host'],
			user=settings['mysql']['username'],
			passwd=settings['mysql']['password'],
			database=settings['mysql']['database']
		)
		print colored('SUCCESS', 'green')
	except Exception as e:
		print colored('ERROR', 'red')
		print colored(e, 'red')
		exit()

# Variables
queue = []
lastread = 0

# Open a subprocess and listen to the radio
multimon_ng = subprocess.Popen(settings['radio']['command'], stdout=subprocess.PIPE, stderr=open(os.devnull, 'w'), shell=True)

# Make subprocess non-blocking so we can keep track of other stuff while nothing is happening
fcntl.fcntl(multimon_ng.stdout.fileno(), fcntl.F_SETFL, os.O_NONBLOCK)

# Check if we can access radio
time.sleep(0.5)
if multimon_ng.poll() != None:
	print colored('Cannot claim radio thread, attempting override...', 'magenta')
	killRadio()
	time.sleep(1)
	multimon_ng = subprocess.Popen(settings['radio']['command'], stdout=subprocess.PIPE, stderr=open(os.devnull, 'w'), shell=True)
	fcntl.fcntl(multimon_ng.stdout.fileno(), fcntl.F_SETFL, os.O_NONBLOCK)
	if multimon_ng.poll() != None:
		print colored('Could not force override, terminating...', 'red')
		sys.exit()
	else:
		print colored('Reclaimed radio, proceeding...', 'green')

# Initialization info
print 'Started listener with PID:',
print colored(multimon_ng.pid, 'cyan')
print colored("Watching...", 'cyan')

try:
	# We wanna run another cycle?
    while True:
		time.sleep(0.0001) # Add a sleep so we don't burn our CPU (oops)
		line = '' # Store the received line

		# Check accessibility of the radio
		if multimon_ng.poll() != None:
			print colored('Radio terminated by external control.', 'red')
			break

		# Readline from radio
		try:
			line = multimon_ng.stdout.readline()
			saveraw(line, settings)
			if line.__contains__("ALN") and line.startswith('FLEX'):
				queue.append(line)
				lastread = time.time()
		
		# Nothing to read...
		except:
			pass

		# Process queued messages after triggertime
		if time.time() - lastread > settings['radio']['triggertime'] and len(queue) > 0:
			last_msg = ""
			last_line = ""
			capcodes = []

			cnt = 1
			queue.append('X' * 64)
			for line in queue:
				# Substring the needed parts and set globals
				capcode = line[47:54]
				message = line[60:]

				# If first or same msg, append capcode
				if message == last_msg or cnt == 1:
					last_msg = message
					last_line = line
					capcodes.append(capcode)
				
				# If different msg or last queue item, push out alert
				if message != last_msg:
					# Save raw if wanted
					saverawunique(last_msg, settings)

					capinfo = getCapInfo(settings, capcodes)
					prio = getPrio(last_msg)

					printMessage(last_line, prio)
					printCapInfo(settings, capinfo, capcodes)

					msgobject = {
						"timestamp": last_line[6:25],
						"message": last_msg,
						"capcodes": capcodes,
						"capinfo": capinfo,
						"prio": prio
					}

					process(settings, msgobject)

					# Set for next
					last_msg = message
					last_line = line
					capcodes = [capcode]
				cnt += 1
			queue = []

# Keyboard Interrupt (Ctrl + C)
except KeyboardInterrupt:
	print colored('\nTerminated by user.', 'red')

# Cleanup
finally:
	killRadio()
	if con:
		con.close()